File: statefun-python-sdk/statefun/storage.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

import typing
from dataclasses import dataclass

from statefun.core import ValueSpec, Type
from statefun.request_reply_pb2 import TypedValue, ToFunction

from statefun.utils import to_typed_value, from_typed_value


@dataclass(frozen=True, repr=True, eq=True, order=False)
class StorageSpec:
    make_instance: typing.Any
    names: typing.FrozenSet[str]
    specs: typing.List[ValueSpec]


@dataclass(frozen=True, eq=False, repr=False, order=False)
class Resolution:
    missing_specs: typing.Union[None, typing.List[ValueSpec]]
    storage: typing.Union[None, typing.Any]  # this holds the storage instance.


class Cell(object):
    __slots__ = ("tpe", "typed_value", "dirty")

    def __init__(self, tpe: Type, typed_value: typing.Optional[TypedValue]):
        # read only
        self.tpe = tpe
        # mutable
        self.typed_value = typed_value
        self.dirty = False

    def get(self):
        typed_value = self.typed_value
        return from_typed_value(self.tpe, typed_value)

    def set(self, val):
        if val is None:
            raise ValueError('provided value must not be None. To delete a value, please use del.')
        tpe = self.tpe
        typed_value = to_typed_value(tpe, val)
        self.typed_value = typed_value
        self.dirty = True

    def delete(self):
        self.typed_value = None
        self.dirty = True


# self.cells: typing.Dict[str, Cell] = {name: Cell(name, tpe, vals[name]) for name, tpe in types.items()}


def storage_constructor(self, cells: typing.Dict[str, Cell]):
    self._cells = cells


def property_named(name):
    """
    Creates a property, that delegates all the operations to an instance
    of a HiddenClass, that is expected to be member of target class (the class that
    this property will be part of).

    :param name: the name of this field
    :return: a property as described above.
    """

    def fget(self):
        cell: Cell = self._cells.get(name)
        if not cell:
            raise AttributeError(name)
        return cell.get()

    def fset(self, val):
        cell: Cell = self._cells.get(name)
        if not cell:
            raise AttributeError(name)
        cell.set(val)

    def fdel(self):
        cell: Cell = self._cells.get(name)
        if not cell:
            raise AttributeError(name)
        cell.delete()

    return property(fget, fset, fdel)


def make_address_storage_spec(specs: typing.List[ValueSpec]) -> StorageSpec:
    """
    Creates an StorageSpec from user supplied value specs.
    :param specs: a list of specs as supplied by the user.
    :return: a StorageSpec.
    """
    props = {"__init__": storage_constructor, "__slots__": ["_cells"]}
    for spec in specs:
        if spec.name in props:
            raise ValueError("duplicate registered value name: " + spec.name)
        props[spec.name] = property_named(spec.name)
    cls = type("GeneratedAddressedScopedStorage", (object,), props)
    return StorageSpec(make_instance=cls,
                       names=frozenset(spec.name for spec in specs),
                       specs=specs)


def resolve(storage: StorageSpec,
            values: typing.List[ToFunction.PersistedValue]) -> Resolution:
    """
    Resolve the registered specs and the actually received values.

    :param storage: a storage factory
    :param values: the actually received values
    :return: a Resolution result, that might have either a list of missing specs
    (specs that were defined by the user but didn't arrived from StateFun) or a
     successful resolution, with an instance of an addressed scoped storage.
    """
    #
    # index the actually received values (TypedValue) by name.
    #
    received: typing.Dict[str, TypedValue] = {state.state_name: state.state_value for state in values}
    #
    # see if any of the specs are missing.
    #
    missing_keys = storage.names - received.keys()
    if missing_keys:
        # keep the missing specs in exactly the same order as they were originally defined.
        # This is not strictly required from the protocol point of view, but it makes it a bit easier
        # to troubleshoot.
        missing = [spec for spec in storage.specs if spec.name in missing_keys]
        return Resolution(missing_specs=missing, storage=None)
    else:
        cells: typing.Dict[str, Cell] = {spec.name: Cell(tpe=spec.type, typed_value=received[spec.name]) for spec in
                                         storage.specs}
        s = storage.make_instance(cells)
        return Resolution(missing_specs=None, storage=s)

File: statefun-sdk-js/src/storage.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";

import "./generated/request-reply_pb";
import {TypedValueSupport} from "./types";
import {Type, ValueSpec} from "./core";

const MutationType = proto.io.statefun.sdk.reqreply.FromFunction.PersistedValueMutation.MutationType;

// noinspection JSValidateJSDoc
export class Value<T> {
    readonly name: string;
    readonly #type: Type<T>;
    #box: proto.io.statefun.sdk.reqreply.TypedValue | null;
    #mutated: boolean;
    #deleted: boolean;

    /**
     *
     * @param {string} name
     * @param {Type} type
     * @param {proto.io.statefun.sdk.reqreply.TypedValue} box
     */
    constructor(name: string, type: Type<T>, box: proto.io.statefun.sdk.reqreply.TypedValue | null) {
        this.name = name;
        this.#type = type;
        this.#box = box;
        this.#mutated = false;
        this.#deleted = false;
    }

    getValue(): T | null {
        if (this.#deleted) {
            return null;
        }
        return TypedValueSupport.parseTypedValue(this.#box, this.#type);
    }

    setValue(jsObject: T | null) {
        if (jsObject === undefined || jsObject === null) {
            this.#mutated = true;
            this.#deleted = true;
            this.#box = null;
        } else {
            this.#mutated = true;
            this.#deleted = false;
            this.#box = TypedValueSupport.toTypedValue(jsObject, this.#type);
        }
    }

    // internal helpers

    asMutation(): proto.io.statefun.sdk.reqreply.FromFunction.PersistedValueMutation | null {
        if (!this.#mutated) {
            return null;
        }
        const mutation = new proto.io.statefun.sdk.reqreply.FromFunction.PersistedValueMutation();
        mutation.setStateName(this.name);
        if (this.#deleted) {
            mutation.setMutationType(MutationType['DELETE']);
        } else {
            mutation.setMutationType(MutationType['MODIFY']);
            mutation.setStateValue(this.#box);
        }
        return mutation;
    }

    static fromState<U>(persistedValue: proto.io.statefun.sdk.reqreply.ToFunction.PersistedValue, type: Type<U>) {
        const name = persistedValue.getStateName();
        return new Value<U>(name, type, persistedValue.getStateValue());
    }
}

// noinspection JSValidateJSDoc
export class AddressScopedStorageFactory {

    /**
     * Tries to create an AddressScopedStorage. An object that contains each known state as a property on that object.
     *
     * @param {proto.io.statefun.sdk.reqreply.ToFunction.InvocationBatchRequest} invocationBatchRequest
     * @param { [ValueSpec] } knownStates
     * @returns either a list of missing ValueSpecs or a list of Values and an AddressScopedStorage.
     */
    static tryCreateAddressScopedStorage(
        invocationBatchRequest: proto.io.statefun.sdk.reqreply.ToFunction.InvocationBatchRequest,
        knownStates: ValueSpec[]
    ) {
        const receivedState = AddressScopedStorageFactory.indexActualState(invocationBatchRequest);
        const {found, missing} = AddressScopedStorageFactory.extractKnownStates(knownStates, receivedState);
        if (missing.length > 0) {
            // the caller needs to respond with an IncompleteInvocationResponse,
            // that contains all the missing specs.
            return {
                missing: missing,
                values: null,
                storage: null
            }
        }
        // TODO: consider caching and setting the newly received states by calling setValue on each individual value.
        const storage = AddressScopedStorageFactory.create(found);
        return {
            missing: null,
            values: found,
            storage: storage,
        };
    }

    static extractKnownStates(knownStates: ValueSpec[], receivedState: Record<string, any>) {
        const found = [];
        const missing = [];
        for (const spec of knownStates) {
            if (!receivedState.hasOwnProperty(spec.name)) {
                missing.push(spec);
                continue;
            }
            const persistedValue = receivedState[spec.name];
            found.push(Value.fromState(persistedValue, spec.type));
        }
        return {found, missing};
    }

    static indexActualState(
        batchRequest: proto.io.statefun.sdk.reqreply.ToFunction.InvocationBatchRequest
    ): Record<string, any> {
        const states = batchRequest.getStateList();
        const gotState: Record<string, any> = {};
        for (const state of states) {
            gotState[state.getStateName()] = state;
        }
        return gotState;
    }

    /**
     * @param {[Value]} values a list of initialize values
     */
    static create(values: Value<unknown>[]) {
        const storage = Object.create(null);
        for (const v of values) {
            Object.defineProperty(storage, v.name, {
                get: () => v.getValue(),
                set: (newValue) => v.setValue(newValue)
            })
        }
        return Object.seal(storage);
    }

    static collectMutations(values: Value<unknown>[]): proto.io.statefun.sdk.reqreply.FromFunction.PersistedValueMutation[] {
        return <proto.io.statefun.sdk.reqreply.FromFunction.PersistedValueMutation[]>values
            .map(value => value.asMutation())
            .filter(mutation => mutation !== null);
    }
}
